### ISSUE TITLE
No attribute 'outer_context' when calculating gradient from imported graph

### ISSUE THREAD

**Issue by malmaud** [tensorflow/tensorflow#7404]
It seems when you import a graph with a "while" loop, you can't calculate gradients as you could on the original graph. e.g.

```python
import tensorflow as tf
i=tf.constant(0, name="input")
out=tf.while_loop(lambda i: tf.less(i,5), lambda i: [tf.add(i,1)], [i], name="output")
graph_def = tf.get_default_graph().as_graph_def()

g = tf.Graph()
with g.as_default():
    tf.import_graph_def(graph_def)
s = tf.Session(graph=g)
i_imported = g.get_tensor_by_name("import/input:0")
out_imported = g.get_tensor_by_name("import/output/Exit:0")
tf.gradients(out_imported, i_imported)
```

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-12-e7e2b78684d3> in <module>()
----> 1 tf.gradients(out_imported, i_imported)

/Users/malmaud/anaconda/lib/python2.7/site-packages/tensorflow/python/ops/gradients_impl.pyc in gradients(ys, xs, grad_ys, name, colocate_gradients_with_ops, gate_gradients, aggregation_method)
    439     pending_count, loop_state = _PendingCount(ops.get_default_graph(), to_ops,
    440                                               from_ops,
--> 441                                               colocate_gradients_with_ops)
    442 
    443     # Iterate over the collected ops.


/Users/malmaud/anaconda/lib/python2.7/site-packages/tensorflow/python/ops/gradients_impl.pyc in _PendingCount(graph, to_ops, from_ops, colocate_gradients_with_ops)
    184   # 'loop_state' is None if there are no while loops.
    185   loop_state = control_flow_ops.MaybeCreateControlFlowState(
--> 186       between_op_list, between_ops, colocate_gradients_with_ops)
    187 
    188   # Initialize pending count for between ops.

/Users/malmaud/anaconda/lib/python2.7/site-packages/tensorflow/python/ops/control_flow_ops.pyc in MaybeCreateControlFlowState(between_op_list, between_ops, colocate_gradients_with_ops)
   1293           loop_state.AddWhileContext(op, between_op_list, between_ops)
   1294       else:
-> 1295         loop_state.AddWhileContext(op, between_op_list, between_ops)
   1296   return loop_state
   1297 

/Users/malmaud/anaconda/lib/python2.7/site-packages/tensorflow/python/ops/control_flow_ops.pyc in AddWhileContext(self, op, between_op_list, between_ops)
   1102     if grad_state is None:
   1103       # This is a new while loop so create a grad state for it.
-> 1104       outer_forward_ctxt = forward_ctxt.outer_context
   1105       if outer_forward_ctxt:
   1106         outer_forward_ctxt = outer_forward_ctxt.GetWhileContext()

AttributeError: 'NoneType' object has no attribute 'outer_context'
```

**Comment by girving** [comment]
@ebrevdo Any idea what would be happening here?  Do while loop gradients depend on unserialized information?

**Comment by ebrevdo** [comment]
Yes, though I thought Yuan and Sherry had added the appropriate serialization to the metagraph.  Is this on master branch of tf?

**Comment by malmaud** [comment]
Ya, this is master.

**Comment by malmaud** [comment]
OK, I see if I save and restore the metagraph (via `tf.train.export_meta_graph` and `tf.train.import_meta_graph`, then I can take the gradient in the restored session. So it's just a problem if you try serializing and importing the `GraphDef` itself.

**Comment by girving** [comment]
@ebrevdo Is there an easy way to produce a nicer error message here, so that users don't have to guess the graph vs. metagraph issue?

**Comment by ebrevdo** [comment]
Yes; I believe that the forward_ctxt object here:

forward_ctxt.outer_context

should have a @property outer_context; and if no such internal object is
set, it can raise a ValueError/InternalError that maybe you're trying to
call gradients on a while loop without properly serializing your graph via
MetaGraphDef.

**Comment by kejiashi** [comment]
Hi @ebrevdo , is there any update on this issue? Thanks!

**Comment by posenhuang** [comment]
@ebrevdo, any update on this issue or any workaround? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by gunan** [comment]
@ebrevdo should we mark this contributions welcome?

**Comment by ebrevdo** [comment]
Yes

**Comment by imsheridan** [comment]
Created a PR #18052 to fix this.

**Comment by Tal-Golan** [comment]
This issue is not really resolved. In some cases, you can't just replace tf.import_graph_def with  tf.train.import_meta_graph. The latter doesn't seem to have all of the functionality of the former. For example, I need to use the input_map and return_elements arguments to connect the loaded graph into an existing graph.

**Comment by ProjectDimlight** [comment]
Hello, I came up with the same issue when I use **tf.import_graph_def**:
AttributeError: 'NoneType' object has no attribute 'outer_context'

Following the instructions above I switched to import_meta_graph like this:
`
    def _make_model_and_ops(self, patch_val):
        start = time.time()
        
        self.image_input_ = tf.placeholder(tf.float32, shape=(None, None, None, 3), name='image_input')
        saver = tf.train.import_meta_graph(PATH_TO_CKPT + '.meta',
                                           import_scope="detection",
                                           input_map={
                                               'ToFloat_3': self.image_input_
                                           })
        saver.restore(self.sess, PATH_TO_CKPT)
        self.graph = self.sess.graph
        
        with self.sess.graph.as_default():
            tf.set_random_seed(1234)
            
            # Tensors are post-fixed with an underscore!
            #self.image_input_shape_ = tf.placeholder(tf.int32, shape=(1,3), name="image_input_shape")
            
            self.eps_ = tf.placeholder(tf.float32, shape=(1), name='eps')
            
            # The following commented part is the original code using import_graph_def
            '''
            detection_graph_def = tf.GraphDef()
            with tf.gfile.GFile(PATH_TO_FROZEN_GRAPH, 'rb') as fid:
                serialized_graph = fid.read()
                detection_graph_def.ParseFromString(serialized_graph)
                tf.import_graph_def(detection_graph_def, name='detection',
                                    input_map={
                                               'Preprocessor/map/TensorArrayStack/TensorArrayGatherV3': self.image_input_,
                                               'Preprocessor/map/TensorArrayStack_1/TensorArrayGatherV3': self.image_input_shape_
                                              })
            '''
            
            # Second-stage Class Loss
            self.second_stage_cls_scores_ = self.graph.get_tensor_by_name('detection/SecondStagePostprocessor/convert_scores:0')
            second_stage_cls_logits_ = self.graph.get_tensor_by_name('detection/SecondStagePostprocessor/scale_logits:0')
            self.second_stage_cls_labels_ = tf.placeholder(tf.float32, shape=second_stage_cls_logits_.shape, name='second_stage_cls_labels')
            
            self.second_stage_cls_losses_ = tf.nn.softmax_cross_entropy_with_logits_v2(labels=tf.reshape(self.second_stage_cls_labels_, (-1, self.second_stage_cls_labels_.shape[2])),
                                                                                       logits=tf.reshape(second_stage_cls_logits_, (-1, second_stage_cls_logits_.shape[2]))) 
           
            # Second-stage bounding boxes
            self.second_stage_loc_bboxes_ = self.graph.get_tensor_by_name('detection/SecondStagePostprocessor/Reshape_4:0')
            
            grads = tf.gradients(self.second_stage_cls_losses_, [self.image_input_])
            print(grads)
            self.unclipped_adv_images_ = self.image_input_ + self.eps_ * tf.sign(grads)
            self.adv_images_ = tf.clip_by_value(self.unclipped_adv_images_, clip_value_min=0, clip_value_max=255)
    
    
        elapsed = time.time() - start
        print("Finished loading the model, took {:.0f}s".format(elapsed))           
`

But the problem still exist.
Does this mean that the information required for tf.gradients does not appear in the check point? If so, what else can I do?

Thanks, and sorry for my poor English....

**Comment by SilverKnightVGM** [comment]
Having the same problem as @ProjectDimlight, tried the suggestions above with no luck. Any updates on this?
